Blazor.ExtraDry/Blazor.ExtraDry.Core/Rules/RuleEngine.cs

```diff
@@ -138,36 +138,91 @@ private async Task<object> ResolveEntityValue(PropertyInfo property, object sour
+        /// <param name="hardDeletePrepare">If item can't be soft-deleted, then the action that is executed for a hard-delete.</param>
+        /// <param name="hardDeleteCommit">If item can't be soft-deleted, then the optional action that is executed to commit hard-delete.</param>
         [SuppressMessage("Performance", "CA1822:Mark members as static", Justification = "Keep as standard service instance style for DI.")]
-        public void Delete<T>(T item, Action hardDeleteOption = null)
+        public void DeleteSoft<T>(T item, Action hardDeletePrepare, Action hardDeleteCommit)
         {
             if(item == null) {
                 throw new ArgumentNullException(nameof(item));
             }
-            var properties = typeof(T).GetProperties();
+            if(!AttemptSoftDelete(item)) {
+                if(hardDeletePrepare == null) {
+                    throw new InvalidOperationException("Item could not be soft-deleted and no hard delete action was provided.");
+                }
+                else {
+                    hardDeletePrepare();
+                    hardDeleteCommit?.Invoke();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Processes a hard delete of an item if possible, even if a `DeleteValue` is provided in the `RulesAttribute`.
+        /// This allows for the database to notify if a hard-delete is not possible by throwing an exception in the commit.
+        /// If not possible, then the soft-delete is performed instead.
+        /// If neither is possible, an exception is thrown.
+        /// </summary>
+        /// <param name="item">The item to delete, a soft-delete is attempted first.</param>
+        /// <param name="hardDeletePrepare">If item can't be soft-deleted, then the action that is executed for a hard-delete.</param>
+        /// <param name="hardDeleteCommit">If item can't be soft-deleted, then the optional action that is executed to commit hard-delete.</param>
+        [SuppressMessage("Performance", "CA1822:Mark members as static", Justification = "Keep as standard service instance style for DI.")]
+        public void DeleteHard<T>(T item, Action hardDeletePrepare, Action hardDeleteCommit)
+        {
+            if(item == null) {
+                throw new ArgumentNullException(nameof(item));
+            }
+            if(hardDeletePrepare == null) {
+                throw new ArgumentNullException(nameof(item), "Must provide an action to prepare hard delete, e.g. remove from collection");
+            }
+            if(hardDeleteCommit == null) {
+                throw new ArgumentNullException(nameof(item), "Must provide an action to commit hard delete, e.g. save to database");
+            }
+            try {
+                hardDeletePrepare();
+                hardDeleteCommit();
+            }
+            catch { // Hmmm, can we be specific about exception type here?
```

Comment:
Catching specifics would be ideal, I assume there's some data storage exceptions coming back, but that's all I'd expect in a normal use case.
There could be all sorts of business rules in the prepare and commit though, so to cater for that, I don't think we can be more specific

Reply:
Yep, that's why I left it hanging there...  Not 100% happy, but might have to suck it...